content/backend/graphql-scala/5-extending_link_model.md

```diff
@@ -0,0 +1,154 @@
+
+<Instruction>
+
+Add `createdAt` field in `Link` models for populated example data in `DBSchema`. Change `databaseSetup` function into following code:
+
+```scala
+Links ++= Seq(
+      Link(1, "http://howtographql.com", "Awesome community driven GraphQL tutorial", DateTime(2017,9,12)),
+      Link(2, "http://graphql.org", "Official GraphQL webpage",DateTime(2017,10,1)),
+      Link(3, "https://facebook.github.io/graphql/", "GraphQL specification",DateTime(2017,10,2))
+    )
+```
+
+</Instruction>
+
+But `H2` doesn't know how to store such type in database, we have to help understand this as well.
+
+<Instruction>
+
+Add column mapper for our `DateTime` type in `DBSchema` object.
+
+```scala
+implicit val dateTimeColumnType = MappedColumnType.base[DateTime, Timestamp](
+    dt => new Timestamp(dt.clicks),
+    ts => DateTime(ts.getTime)
+)
+```
+
+</Instruction>
+
+This mapper will convert `DateTime` into used internally by database `Long` type.
+
+The last thing is to add `createdAt` column definition in table declaration.
+
+<Instruction>
+
+Add following code inside `LinksTable` class. Remove current `*` function.
+
+```scala
+def createdAt = column[DateTime]("CREATED_AT")
+
+def * = (id, url, description, createdAt) <> ((Link.apply _).tupled, Link.unapply)
+```
+
+</Instruction>
+
+
+### Define custom scalar for DateTime
+
+Sangria supports all standard GraphQL scalars like `String`, `Int`, etc. In addition you can find scalars for types like `Long`, `BigInt` or `BigDecimal`. There are a lot of them actually. But, probably it still too less to cover all your needs.
+Like in our example, we're using `DateTime` type, and there are no built-in scalar for such type.
+We have to make our own to make able present it as string (and eventually take back from string input and make auto conversion to our type).
+
+Let's write a scalar that converts `String` and `DateTime`. In both ways.
+
+<Instruction>
+
+In `GraphQLSchema` file add following code:
+
+```scala
+implicit val GraphQLDateTime = ScalarType[DateTime](//1
+  "DateTime",//2
+  coerceOutput = (dt, _) => dt.toString, //3
+  coerceInput = { //4
+    case StringValue(dt, _,_ ) => DateTime.fromIsoDateTimeString(dt).toRight(DateTimeCoerceViolation)
+    case _ => Left(DateTimeCoerceViolation)
+  },
+  coerceUserInput = { //5
+    case s: String => DateTime.fromIsoDateTimeString(s).toRight(DateTimeCoerceViolation)
+    case _ => Left(DateTimeCoerceViolation)
+  }
+)
+
+```
+</Instruction>
+
+1. Use `implicit` because it's implicitly has to be in scope
+1. `"DateTime"` is a name to this scalar. It will be visible in schema with that name.
+1. `coerceOutput` converts our type to string. It will be used to produce output data.
+1. `coerceInput` needs partial function with `Value` as single argument. Such value could be of many types. In our case we're parsing only from `StringValue` but you can also add `LongType` to able converts from timestamp, so user will have option what to choose.
```

Comment:
`... from StringValue. Of course nothing stops you from defining few conversions. If you define more cases for coerceInput users will have freedom to provide input in more ways`